gnomad/utils/constraint.py

```diff
@@ -360,3 +370,254 @@ def collapse_strand(
+        low_coverage_obs_exp=hl.agg.sum(all_low_coverage_ht.observed_variants)
+        / (
+            high_coverage_scale_factor
+            * hl.agg.sum(
+                all_low_coverage_ht.possible_variants * all_low_coverage_ht.mu_snp
+            )
+        )
+    )
+
+    # Build the coverage model
+    # TODO: consider weighting here as well
+    coverage_model_expr = build_coverage_model(
+        low_coverage_obs_exp=low_coverage_ht.low_coverage_obs_exp,
+        log_coverage=low_coverage_ht.log_coverage,
+    )
+    coverage_model = tuple(low_coverage_ht.aggregate(coverage_model_expr).beta)
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
+    cpg_expr: hl.expr.BooleanExpression,
+    mu_snp_expr: hl.expr.Float64Expression,
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_exprs: List[hl.ArrayNumericExpression] = [],
+    weighted: bool = False,
+) -> Dict[str, Dict[bool, hl.expr.ArrayExpression]]:
+    """
+    Build plateau models for all the sites in `ht` and sites in each `pop` population downsampling to calibrate mutation rate to compute predicted proportion observed value.
+
+    The x and y of the plateau models:
+        - x: `mu_snp` - mutation rate
+        - y: proportion observed ('observed_variants' or 'observed_{pop}' / 'possible_variants')
+
+    :param cpg_expr: The BooleanExpression noting whether a site is a CPG site.
+    :param mu_snp_expr: The Float64Expression of the mutation rate.
+    :param observed_variants_expr: The Int64Expression of the observed variant counts for each combination of keys in `ht`.
+    :param possible_variants_expr: The Int64Expression of the possible variant counts for each combination of keys in `ht`.
+    :param pop_observed_variants_exprs: List of ArrayNumericExpression of observed variant counts for specified populations. Default is [].
+    :param weighted: Whether to generalize the model to weighted least squares using 'possible_variants'. Default is False.
+    :return: A Dictionary of intercepts and slopes for plateau models of each population. The key of the dictionary is population name, and the value is a dictionary (or a list of dictionary if `pop_observed_variants_exprs` is specified) mapping cpg BooleanExpression to a intercept and a slope.
+    """
+    # Build a plateau model using all the sites in the Table
+    plateau_models_agg_expr = {
+        "total": hl.agg.group_by(
+            cpg_expr,
+            hl.agg.linreg(
+                observed_variants_expr / possible_variants_expr,
+                [1, mu_snp_expr],
+                weight=possible_variants_expr if weighted else None,
+            ).beta,
+        )
+    }
+    # Build plateau models using sites in population downsamplings if population is specified
+    for pop, pop_observed_variants_expr in pop_observed_variants_exprs.items():
+        plateau_models_agg_expr[pop] = hl.agg.array_agg(
```

Comment:
Konrad [suggested](https://the-tgg.slack.com/archives/C03PP8Y98LA/p1665262483906149?thread_ts=1665177762.851659&cid=C03PP8Y98LA) that we can used `array_agg()` so I add it here. I tested it and it works as expected!!! Just want to double check if it can be simplified.

Reply:
I tried to use `map_value()` and got this error: `dynamic variables created by 'hl.bind' or lambda methods like 'hl.map' may not be aggregated`. I guess it's saying the `array_agg` can't aggregate the dict values given by lambda in `map_value()`, so I kept the old version.

Reply:
ah, works for me. I don't love the for loop + hail expression combination though, maybe you could post to zulip to see if the hail team has any suggestions for how to streamline this?